### ISSUE TITLE
Incorrect merging of texture data.

### ISSUE THREAD

**Issue by serkkz** [panda3d/panda3d#1047]
When exporting an object with different values of the same texture. Panda3D merge them into one by file name, and the rest will be lost.

Test egg:
```
<CoordinateSystem> { Z-Up }

<Texture> Texture1 {
 "panda.jpg"
 <Scalar> format { RGB }
 <Scalar> alpha { OFF }
 <Scalar> compression { OFF }
 <Scalar> envtype { MODULATE }
 <Scalar> anisotropic-degree { 16 }
 <Scalar> wrapu { REPEAT }
 <Scalar> wrapv { REPEAT }
 <Scalar> minfilter { LINEAR_MIPMAP_LINEAR }
 <Scalar> magfilter { LINEAR_MIPMAP_LINEAR }
}

<Texture> Texture2 {
 "panda.jpg"
 <Scalar> format { RGB }
 <Scalar> alpha { ON }
 <Scalar> compression { DXT5 }
 <Scalar> envtype { MODULATE }
 <Scalar> anisotropic-degree { 0 }
 <Scalar> wrapu { REPEAT }
 <Scalar> wrapv { BORDER_COLOR }
 <Scalar> minfilter { LINEAR_MIPMAP_LINEAR }
 <Scalar> magfilter { LINEAR_MIPMAP_LINEAR }
}

<Group>  Plane {
 <VertexPool> Plane {
  <Vertex> 0 { -1 -1 0 
  <UV> { 0 0 }
  }
  <Vertex> 1 { 1 -1 0 
  <UV> { 1 0 }
  }
  <Vertex> 2 { 1 1 0 
  <UV> { 1 1 }
  }
  <Vertex> 3 { -1 1 0 
  <UV> { 0 1 }
  }
  <Vertex> 4 { -1 1.40914 0 
  <UV> { 0 0 }
  }
  <Vertex> 5 { 1 1.40914 0 
  <UV> { 1 0 }
  }
  <Vertex> 6 { 1 3.40914 0 
  <UV> { 1 1 }
  }
  <Vertex> 7 { -1 3.40914 0 
  <UV> { 0 1 }
  }
 }

 <Polygon> 0 { 
  <TRef> { Texture1}
  <Normal> { 0 0 1 }
  <VertexRef> { 0 1 2 3 <Ref> { Plane } }
  }
  
 <Polygon> 1 { 
  <TRef> { Texture2 }
  <Normal> { 0 0 1 }
  <VertexRef> { 4 5 6 7 <Ref> { Plane } }
  }
}
```
app:

```py
from direct.showbase.ShowBase import ShowBase
from panda3d.core import TextureAttrib

base = ShowBase()

musicBox = loader.loadModel('plane')
result = musicBox.find('**/Plane')

Geom0 = result.node().getGeomState(0)
texture0 = Geom0.getAttrib(TextureAttrib).getTexture()

Geom1 = result.node().getGeomState(1)
texture1 = Geom1.getAttrib(TextureAttrib).getTexture()

print(texture0)
print(texture1)

base.run()
```
print ouput:

```bat
Known pipe types:
  wglGraphicsPipe
(all display modules loaded.)
2d_texture panda (from /f/Code/Panda3D/Python/MaterialEditor/Test/panda.jpg)
  2-d, 512 x 512 pixels, each 3 bytes, rgb, compression dxt5
  sampler wrap(u=repeat, v=border_color, w=repeat, border=0 0 0 1) filter(min=linear_mipmap_linear, mag=linear, aniso=0) lod(min=-1000, max=1000, bias=0)  262144 bytes in ram, compression dxt5
  9 mipmap levels also present in ram (87408 bytes).
  simple image: 16 x 16, 1024 bytes

2d_texture panda (from /f/Code/Panda3D/Python/MaterialEditor/Test/panda.jpg)
  2-d, 512 x 512 pixels, each 3 bytes, rgb, compression dxt5
  sampler wrap(u=repeat, v=border_color, w=repeat, border=0 0 0 1) filter(min=linear_mipmap_linear, mag=linear, aniso=0) lod(min=-1000, max=1000, bias=0)  262144 bytes in ram, compression dxt5
  9 mipmap levels also present in ram (87408 bytes).
  simple image: 16 x 16, 1024 bytes
```
![panda](https://user-images.githubusercontent.com/17797036/98453417-cf694f00-2169-11eb-8c5a-b503102d77f8.jpg)

Windows 10 x64
Panda3D SDK 1.10.7
Python 3.7.9

**Comment by rdb** [comment]
We will need to add a check to EggLoader::load_texture to make a copy of the Texture when the properties differ.

**Comment by serkkz** [comment]
In fact, I think it's easier to compare the texture tag name.

```html
<Texture> Texture1 { }
```
and

```html
<Texture> Texture2 { }
```
Result: these are different textures.

**Comment by rdb** [comment]
That is not sufficient.  You may want to apply the same texture to two different texture stages with different envtype, without causing a duplication of the texture.  We do not want to introduce a performance regression.

It's also not sufficient when there are two egg files referencing the same texture, but with the same properties.  These should not be duplicated simply because the names are different.

**Comment by serkkz** [comment]
However, the model loader does not look at the parameters of loaded models. you can load the same model with a different name, and this is correct. What you are trying to optimize should actually be on the users ' shoulders. they should not duplicate the same texture description under different names. In fact, the texture pool must decide whether to load a new texture or return a link if it was loaded with the same name.

The most interesting thing is that this problem will go away if you introduce the concept of a library. Material and Texture, only this needs to be combined, under the concept of material.

**Comment by rdb** [comment]
In reality, a <Texture> in an egg file doesn't actually correspond 1:1 with a Texture in Panda3D, and furthermore, the name of the <Texture> is often not really in control by the user—it's synthesized by the converter, or it's something silly from the modelling program like "Texture.001".  I don't think we should perform important logic based on something so arbitrary.  If we wanted to do it this way, we should have made that part of the specification of the .egg format long ago.  Unfortunately, with egg being what it is, we are already too far down the route of having to deduplicate texture objects.

**Comment by paustian** [comment]
I dug into this one a little. And I think I found the code that needs to be changed. In [EggTextureCollection::
find_used_textures](https://github.com/panda3d/panda3d/blob/79714392cbb475e1a52a94156f70d8c9705c78d6/panda/src/egg/eggTextureCollection.cxx#L169) it compares the keys for the textures, which I bet is their file names. A different find routine would need to be written that compares the properties. Is that correct?

**Comment by paustian** [comment]
I dug into this one a little. And I think I found the code that needs to be changed. In [EggTextureCollection::
find_used_textures](https://github.com/panda3d/panda3d/blob/79714392cbb475e1a52a94156f70d8c9705c78d6/panda/src/egg/eggTextureCollection.cxx#L169) it compares the keys for the textures, which I bet is their file names. A different find routine would need to be written that compares the properties. Is that correct?

**Comment by paustian** [comment]
An update. I got a sample program to demonstrate the issue and have tracked it down to loadSync, but I cannot dig deeper for some reason. I did notice that the sample program had to be slightly different that what was listed above. Here is the sample that demosntrated the issue:

**Comment by paustian** [comment]
With some more digging, I think I found the code that needs to be changed at [TexturePool::
ns_load_texture](https://github.com/panda3d/panda3d/blob/2156e6deb2dc5ecdfe57e2f74c39024075645d62/panda/src/gobj/texturePool.cxx#L458) Before I make an attempt, I want to make sure I am right. At present textures are kept in a map object using this struct as the key:
```
struct LookupKey {
    Filename _fullpath;
    Filename _alpha_fullpath;
    int _primary_file_num_channels = 0;
    int _alpha_file_channel = 0;
    Texture::TextureType _texture_type = Texture::TT_2d_texture;

    INLINE bool operator < (const LookupKey &other) const;
  };
```
Since none of these parameters are unique between the two textures, it grabs the first match. Could the fix be to simply add more information to LookupKey? I find it interesting that it matches texture2 and not texture1. Thoughts?

**Comment by rdb** [comment]
@paustian Thanks for looking into this!

LookupKey in TexturePool contains all the parameters that someone might pass to `TexturePool::load_texture`, however, this does not include all the things that can be specified in an egg file.

One solution might be to have all these egg `<Texture>` options that directly affect the Texture object (read: are assigned by `EggLoader::load_texture()` and `EggLoader::apply_texture_attributes()`) be specified in the LoaderOptions object that is passed to `load_texture`, so that they can become part of the LookupKey.

The alternative solution is to make `EggLoader::load_texture()` aware when a texture has already been loaded with a different set of properties, and copy the returned texture object as needed.  This may just require adding the Texture pointer to a set or something, but you'd have to look into this.

**Comment by paustian** [comment]
After digging through this a bit more, I think the first solution is the one I would want to use. As a bonus, I can reimplement TextturePool to use eggTextureCollection instead of using a pmap<LookupKey, PT(Texture)> Textures. This I think has all the options we are looking for, and as a bonus, already implements a search function for comparing textures. Thoughts.

**Comment by serkkz** [comment]
My thoughts were that the structure when loading from the model file did not differ from the programmatically created one.
If it fixes the situation, then why not. However, rdb does not want to change this behavior, so as not to break user projects. 

Perhaps I should create this model programmatically so that you can test and identify the difference.

**Comment by paustian** [comment]
I could be wrong, but in looking at the code, only certain things are recorded for comparison, and not all the texture data. What has to happen is the programmatically created one needs to reflect all the texture information so that the code doesn't stomp on textures that are different, yet are attached to the same model. I am going to try to explore this today. 

If you want to create a programmatic example that could make my life easier. Thanks!

**Comment by serkkz** [comment]
There are changes. I wrote you an example and found that the problem has a different look.

It looks strange.

```py
from panda3d.core import Texture
from direct.showbase.ShowBase import ShowBase

class MyApp(ShowBase):

    def __init__(self):
        ShowBase.__init__(self)

        tex1 = loader.loadTexture('panda.jpg')
        tex1.setMagfilter(Texture.FTNearest)
        tex1.setMinfilter(Texture.FTNearest)

        tex2 = loader.loadTexture('panda.jpg')
        tex2.setMagfilter(Texture.FTLinear)
        tex2.setMinfilter(Texture.FTLinear)

        print(tex1)
        print(tex2)

app = MyApp()
app.run()
```
out:

```cmd
2d_texture panda (from panda.jpg)
  2-d, 512 x 512 pixels, each 3 bytes, rgb
  sampler wrap(u=repeat, v=repeat, w=repeat, border=0 0 0 1) filter(min=linear, mag=linear, aniso=0) lod(min=-1000, max=1000, bias=0)  786432 bytes in ram, compression off

2d_texture panda (from panda.jpg)
  2-d, 512 x 512 pixels, each 3 bytes, rgb
  sampler wrap(u=repeat, v=repeat, w=repeat, border=0 0 0 1) filter(min=linear, mag=linear, aniso=0) lod(min=-1000, max=1000, bias=0)  786432 bytes in ram, compression off
```

It seems that there is undefined behavior in the texture pool mechanism.

**Comment by serkkz** [comment]
I found that the BAM created through the code has this problem, it may be worth looking in a different direction.

**Comment by serkkz** [comment]
It turned out that when loading BAM, TexturePool actually participates, which is actually strange. If that's the case, we're on the right track.  Sorry for the rush.

**Comment by paustian** [comment]
I have begun the implementation of a solution and I wanted someone with more experience to look at it before I procedure too far down a path. Basically, I am expanding the LoaderOptions class to carry with it all the data necessary to do comparisons of textures and ensure that all unique textures are remembered. 

The [changes](https://github.com/paustian/panda3d/commit/54367fd132fe74d13691d6e4a52c83788b15e12e) appear in EggLoader.cxx, loaderOptions.h and loaderOptions.I. Thanks for any feedback.

**Comment by paustian** [comment]
I think I have a good solution for this issue. (Note that I am writing such a long explanation more for me than anyone else) I was able to change loaderOptions such that the modifications detect the format, compression, and wrap parameters of all the textures that are loaded and can compare values between them. The implementation was able to successfully differentiate the textures in the test code and print out the parameters correctly. Here is a brief summary of how this was implemented:

Textures for models are stored in texturePool. These are kept in a map template, with the key differentiating them. Previously, the LookupKey only differentiated textures by their file path and the number of channels. Thus, if a file had more than one texture in it, the second would get ignored. To solves this I added Bit switches to the LookupKey for each scalar (so far I have implemented format, compression, and wrap scalars). These are then added to LookupKey as thus:
```
struct LookupKey {
    Filename _fullpath;
    Filename _alpha_fullpath;
    int _primary_file_num_channels = 0;
    int _alpha_file_channel = 0;
    int _texture_format = 0;
    int _texture_format_compress = 0;
    int _wrap_options_ou = 0;
    int _wrap_options_vw = 0;
    Texture::TextureType _texture_type = Texture::TT_2d_texture;

    INLINE bool operator < (const LookupKey &other) const;
  };
```
When the .egg file is read, the textures are parsed into an EggTexture.  During load_texture all the options can be explored and the bit switches in LookupKey set. The final piece of the implementation is to change the inline < operator such that it also checks the bit switches when comparing textures. 

My one question left is, how many of the scalar options on the texture listed in the[ panda3d manual](https://docs.panda3d.org/1.10/python/tools/model-export/egg-syntax) need to be implemented?
